gprofiler/perf.py

```diff
@@ -2,77 +2,139 @@
 # Copyright (c) Granulate. All rights reserved.
 # Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
 #
-import concurrent.futures
 import os
-from tempfile import NamedTemporaryFile
+import signal
 from threading import Event
-from typing import Optional, Tuple
+from typing import List, Optional
 
 import psutil
 
+from gprofiler.exceptions import StopEventSetException
 from gprofiler.log import get_logger_adapter
 from gprofiler.merge import ProcessToStackSampleCounters, merge_global_perfs
 from gprofiler.profiler_base import ProfilerBase
-from gprofiler.utils import TEMPORARY_STORAGE_PATH, resource_path, run_process
+from gprofiler.utils import resource_path, run_process, start_process, wait_event, wait_for_file_by_prefix
 
 logger = get_logger_adapter(__name__)
 
-PERF_BUILDID_DIR = os.path.join(TEMPORARY_STORAGE_PATH, "perf-buildids")
+
+class PerfProcess:
+    _dump_timeout_s = 5
+    _poll_timeout_s = 5
+
+    def __init__(self, frequency: int, stop_event: Event, output_path: str, is_dwarf: bool, extra_args: List[str]):
+        self._frequency = frequency
+        self._stop_event = stop_event
+        self._output_path = output_path
+        self._extra_args = extra_args
+        self._type = "dwarf" if is_dwarf else "fp"
+        self._process: Optional[psutil.Process] = None
+
+    def _get_perf_cmd(self) -> List[str]:
+        return [
+            resource_path("perf"),
+            "record",
+            "-F",
+            str(self._frequency),
+            "-a",
+            "-g",
+            "-o",
+            self._output_path,
+            "--switch-output=signal",
+        ] + self._extra_args
+
+    def start(self) -> None:
+        logger.info(f"Starting perf ({self._type} mode)")
+        process = start_process(self._get_perf_cmd(), via_staticx=False)
+        try:
+            wait_event(self._poll_timeout_s, self._stop_event, lambda: os.path.exists(self._output_path))
+        except TimeoutError:
+            process.kill()
+            assert process.stdout is not None and process.stderr is not None
+            logger.error(f"perf failed to start. stdout {process.stdout.read()!r} stderr {process.stderr.read()!r}")
+            raise
+        else:
+            self._process = process
+            logger.info(f"Started perf ({self._type} mode)")
+
+    def stop(self) -> None:
+        if self._process is not None:
+            self._process.terminate()  # okay to call even if process is already dead
+            self._process.wait()
```

Comment:
timeout?

Reply:
`terminate` sends `SIGTERM`, which kills `perf`.

But I agree timeouts are nice nonetheless

Reply:
But we'll add timeouts somewhen else (we have plenty of other missing timeouts - for example, when running py-spy, I happened to see it hang). So in another PR. There's an issue open (or I will open one)